### DISCUSSION TITLE
Module Federation and multiple entries

### DISCUSSION THREAD

**Discussion by meskill** [webpack/webpack#14985] [Q&A]
I have an app with the following setup:
- it has multiple entries. e.g. one is used as bare polyfill entry
- is uses ModuleFederationPlugin with the `react` set as shared eager library

Currently, with that configuration `react` code is bundled in every entry chunk.

Is there any way to prevent ModuleFederation from adding container runtime for every entry or a specific entry as having `react` in polyfill entry is a big overhead.

Made a simple repro to show the issue - https://github.com/meskill/webpack-module-federation-multiple-entries

**Comment by meskill** [comment]
@ScriptedAlchemy looking for a help

**Reply by ScriptedAlchemy** [reply]
From experience with this. Don't set runtime chunk to single in optimization. Or stop using multiple entry points. 

Use remote API from webpack and call the "entry" you want, then execute it. Federation can replace the need for more than one entry. Many of my apps have an empty entry point and exclusively use federation API to kickstart parts of the build as needed.

**Reply by meskill** [reply]
I do not use any runtime chunk options.

I have one additional entry for polyfills which is loaded synchronously only for legacy browsers. As polyfill code should be run synchronously and before any other code I've put it to separate independent entry with its own webpack runtime.

As this entry has such a specific use-case I do not think  I can use webpack runtime as it leads to asynchronous loading and with ModuleFederation it is just duplicates libraries with eager option.

Currently, I've implemented a plugin with hack that prevents adding ModuleFederation on the unwanted entries, but it is pretty ugly inside. Maybe we can add some options for the ModuleFederation plugin in order to allow users to add federation runtime only for specific entries? I can try to contribute to this functionality if this is an acceptable solution.

**Reply by ScriptedAlchemy** [reply]
What about top level await for the polyfill in your index.js file. Then you can await the polyfill before any code is executed or resolve the promise to nothing of not required

**Reply by meskill** [reply]
As I understand, webpack needs a Promise to be available in the environment in order to asynchronously load anything (from [docs](https://webpack.js.org/concepts/#browser-compatibility)) and Promise polyfill will be in that entry. [Webpack docs](https://webpack.js.org/guides/shimming/#loading-polyfills) also suggests to use separate entry for polyfill that can be loaded synchronously

**Reply by meskill** [reply]
I've implemented the [plugin](https://github.com/Tinkoff/tramvai/blob/main/packages/cli/src/library/webpack/plugins/ModuleFederationIgnoreEntries.ts) to enable ignore of specific entries. 

But I think, it might be handled by ModuleFederation itself and might be configured through options, @ScriptedAlchemy wdyt?

**Reply by ScriptedAlchemy** [reply]
So how i work around this right now is by adding dynamic import to all entry points. This fixed it in next.js page routing.

**Reply by zaytsev-mxm** [reply]
Hey @meskill, thanks for the plugin, I've tried that on my build configuration, and it works just fine for me as well!

@ScriptedAlchemy, could please consider introducing an approach similar to one proposed by @meskill in their plugin and probably make it part of the official ModuleFederation API? I am asking because I do agree with you that adding dynamic imports could help — but not for every case. For example, in my case, I use ModuleFederation on a very (very!) old codebase which requires having multiple entries exactly in the way in which they exist now. And ModuleFederation is actually our way for the smooth migration of our legacy to something more new and sweet — part by part. So we simply can not use that approach with migrating all the "old" entries to dynamic imports, even when we really have wanted to.

**Reply by moloneyn** [reply]
Has anyone in this thread found any other solution to this problem or is the plugin still the best option? We're running into the same thing as you @zaytsev-mxm with a legacy code base, but we're hesitant to roll the plugin out to production without it being a fully supported feature and having a better understanding of what the risks are

**Reply by Thebarda** [reply]
Thanks @meskill Your plugin works like a charm 👌

**Reply by ScriptedAlchemy** [reply]
Module-federation/enhanced should support multiple entries without an issue. As well as support for single runtime chunks and custom split chunk options.
